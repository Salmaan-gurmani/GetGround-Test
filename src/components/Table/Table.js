import React, { useState, useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import SearchBar from 'material-ui-search-bar';
import { Typography } from '@material-ui/core';
import Box from '@material-ui/core/Box';
import { Link, useParams } from 'react-router-dom';
import Pagination from '@material-ui/lab/Pagination';
import PaginationItem from '@material-ui/lab/PaginationItem';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import Backdrop from '@material-ui/core/Backdrop';
import CircularProgress from '@material-ui/core/CircularProgress';
const USER_PATH = '/user';

const useStyles = makeStyles(theme => ({
    table: {
        minWidth: 1500,
    },
    backdrop: {
        zIndex: theme.zIndex.drawer + 1,
        color: '#fff',
    },
}));

export default function BooksLibrary({
    data,
    handleQueryParameters,
    isLoading,
}) {
    const { pageNumber } = useParams();
    const [page, setPage] = useState(pageNumber ? pageNumber : 1);
    const [rowsPerPage, setRowsPerPage] = useState(20);
    const [searched, setSearched] = useState('');
    const classes = useStyles();

    useEffect(() => {
        const queryParameters = {
            page: Number(page),
            itemsPerPage: rowsPerPage,
            filter: searched,
        };
        handleQueryParameters(queryParameters);
    }, [rowsPerPage, searched, page]);

    const handleChangePagination = (event, value) => {
        setPage(value);
    };
    const handleChangeRowsCount = event => {
        setRowsPerPage(event.target.value);
    };
    return (
        <>
            <Paper className={classes.table}>
                <SearchBar
                    value={searched}
                    onChange={searched => setSearched(searched)}
                    onCancelSearch={() => setSearched('')}
                />
                <Backdrop className={classes.backdrop} open={isLoading}>
                    <CircularProgress color="inherit" />
                </Backdrop>

                <TableContainer>
                    <Table aria-label="simple table">
                        <TableHead>
                            <TableRow>
                                <TableCell>
                                    <Typography variant="body1">ID</Typography>
                                </TableCell>
                                <TableCell>
                                    <Typography variant="body1">
                                        Title
                                    </Typography>
                                </TableCell>
                                <TableCell>
                                    <Typography variant="body1">
                                        Publication Year
                                    </Typography>
                                </TableCell>
                                <TableCell>
                                    <Typography variant="body1">
                                        Publication Country
                                    </Typography>
                                </TableCell>
                                <TableCell>
                                    <Typography variant="body1">
                                        Publication City
                                    </Typography>
                                </TableCell>
                                <TableCell>
                                    <Typography variant="body1">
                                        Pages
                                    </Typography>
                                </TableCell>
                                <TableCell>
                                    <Typography variant="body1">
                                        Author
                                    </Typography>
                                </TableCell>
                            </TableRow>
                        </TableHead>
                        <TableBody>
                            {data &&
                                data.books &&
                                data.books.map(row => (
                                    <TableRow key={row.id}>
                                        <TableCell component="th" scope="row">
                                            {row.id}
                                        </TableCell>
                                        <TableCell>{row.book_title}</TableCell>
                                        <TableCell>
                                            {row.book_publication_year}
                                        </TableCell>
                                        <TableCell>
                                            {row.book_publication_country}
                                        </TableCell>
                                        <TableCell>
                                            {row.book_publication_city}
                                        </TableCell>
                                        <TableCell>{row.book_pages}</TableCell>
                                        <TableCell>{row.book_author}</TableCell>
                                    </TableRow>
                                ))}
                        </TableBody>
                    </Table>
                </TableContainer>

                <Box
                    display="flex"
                    justifyContent="flex-end"
                    flex={1}
                    padding={1}
                    paddingRight={10}
                >
                    <FormControl>
                        <Select
                            labelId="demo-simple-select-label"
                            id="demo-simple-select"
                            value={rowsPerPage}
                            onChange={handleChangeRowsCount}
                        >
                            <MenuItem value={10}>10</MenuItem>
                            <MenuItem value={20}>20</MenuItem>
                            <MenuItem value={30}>30</MenuItem>
                            <MenuItem value={40}>40</MenuItem>
                            <MenuItem value={50}>50</MenuItem>
                        </Select>
                    </FormControl>
                    <Pagination
                        page={Number(pageNumber)}
                        count={
                            data && data.count
                                ? Math.ceil(data.count / rowsPerPage)
                                : 1
                        }
                        shape="rounded"
                        color="primary"
                        showFirstButton
                        showLastButton
                        boundaryCount={2}
                        onChange={handleChangePagination}
                        renderItem={item => (
                            <PaginationItem
                                type={'start-ellipsis'}
                                component={Link}
                                selected
                                to={`${USER_PATH}/${item.page}`}
                                {...item}
                            />
                        )}
                    />
                </Box>
            </Paper>
            <br />
        </>
    );
}
