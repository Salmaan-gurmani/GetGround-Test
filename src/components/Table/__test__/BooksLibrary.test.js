import React from 'react';
import Table, { PaginationItems } from '../BooksLibrary';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useEffect: jest.fn(),
}));
configure({ adapter: new Adapter() });
const defaultPropsTableComponent = {
    loading: false,
    data: {
        books: [
            {
                book_author: ['Ανώνυμος'],
                book_pages: 104,
                book_publication_city: 'Βενετία',
                book_publication_country: 'Ιταλία',
                book_publication_year: 1529,
                book_title: 'Ο Αλέξανδρος ο Μακεδών',
                id: 2086,
            },
            {
                book_author: ['Ανώνυμος'],
                book_pages: 32,
                book_publication_city: 'Βενετία',
                book_publication_country: 'Ιταλία',
                book_publication_year: 1548,
                book_title:
                    'Διήγησις εις τας πράξεις του περιβοήτου στρατηγού των ρωμαίων μεγάλου Βελισαρίου',
                id: 2060,
            },
        ],
        count: 2000,
    },
    handleQueryParameters: jest.fn(),
};
const loadingEnabledProps = {
    loading: true,
    data: {},
    handleQueryParameters: jest.fn(),
};

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({
        pageNumber: 2,
    }),
}));

describe('<Table />', () => {
    it('should render TableComponent with default values', () => {
        const wrapper = shallow(<Table {...defaultPropsTableComponent} />);
        expect(wrapper).toMatchSnapshot();
    });
    it('should render TableComponent with loading', () => {
        const wrapper = shallow(<Table {...loadingEnabledProps} />);
        expect(wrapper).toMatchSnapshot();
    });
});
describe('<PaginationItems />', () => {
    const item = {
        type: 'page',
        page: 2,
        selected: false,
        disabled: false,
        color: 'primary',
        'aria-label': 'Go to page 2',
        shape: 'rounded',
        size: 'medium',
        variant: 'text',
    };
    it('should render PaginationItems with default values', () => {
        const wrapper = shallow(<PaginationItems item={item} />);
        expect(wrapper).toMatchSnapshot();
    });
});
