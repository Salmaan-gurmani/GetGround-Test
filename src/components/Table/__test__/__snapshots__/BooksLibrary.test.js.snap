// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaginationItems /> should render PaginationItems with default values 1`] = `ShallowWrapper {}`;

exports[`<Table /> should render TableComponent with default values 1`] = `ShallowWrapper {}`;

exports[`<Table /> should render TableComponent with loading 1`] = `ShallowWrapper {}`;
